{% extends "_layouts/default-nocontainer.html" %}

{% block custom_head %}

<!-- jQuery UI CSS -->
<link rel="stylesheet" type="text/css"
      href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>

<!-- IGV CSS -->

<!-- jQuery JS -->
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
<script>$.widget.bridge('uitooltip', $.ui.tooltip);</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha256-KXn5puMvxCw+dAYznun+drMdG1IFl3agK0p/pqT9KAo=" crossorigin="anonymous"></script>

<script type="text/javascript" src="//igv.org/web/release/2.6.2/dist/igv.min.js"></script>

{% endblock %}

{% block content %}

<div class="container-fluid px-5">
  <div class="d-flex justify-content-around flex-wrap mt-2 pb-5">
    <!-- Input Card -->
    <div class="col-md-8 col-xl-3 text-bg-light mb-5 p-5 rounded-2 shadow-sm">
      <!-- Input Form -->
      <form>
        <div class="mb-4">
          <label for="speciesSelect" class="form-label">Species:</label>
          <select class="form-select" aria-label="Choose species" id="speciesSelect">
            <option selected>Choose</option>
            <option value="1">C. Elegans</option>
            <option value="2">C. Briggsae</option>
            <option value="3">C. Tropicalis</option>
          </select>
        </div>
        <!-- Gene search -->
        <div class="mb-4">
          <label for="gene-search" class="form-label">Gene Search:</label>
          <div class="dropdown">
            <input type="text" class="form-control dropdown-toggle" id="gene-search" placeholder="For example, trt-1"
              data-bs-toggle="dropdown">
            <div class="dropdown-menu table-responsive p-1 dropdownSelect">
              <div class="text-center p-3" id="loading-search-table" style="display:none;">
                <div class="spinner-border" role="status">
                  <span class="visually-hidden">Loading...</span>
                </div>
              </div>
              <table id="g-search-table" class='table table-hover table-striped' style="display:none;">
                <caption class="visually-hidden">A list of species strains.</caption>
                <thead>
                  <tr>
                    <th><em>C. elegans Gene</em></th>
                    <th>Symbol/ID</th>
                    <th>Species</th>
                    <th>Source</th>
                  </tr>
                </thead>
                <tbody id="orthologs"></tbody>
              </table>
            </div>
          </div>
        </div>
        <!-- /Gene Search -->
        <div class="mb-4">
          <!-- Isotypes -->
          <label for="filter" class="form-label">Isotype (Reference Strain):</label>
          <div class="dropdown">
            <input type="text" class="form-control dropdown-toggle" id="filter" placeholder="For example, JU360"
              data-bs-toggle="dropdown">
            <div class="dropdown-menu table-responsive dropdownSelect p-1">
              <table id='browser_strain_list' class="table table-striped table-hover">
                <caption class="visually-hidden">A list of strains with different file type options.</caption>
                <tbody class="searchable">
                  {% for strain in strain_listing %}
                  <tr>
                    <td>
                      {{ strain.isotype }} ({{ strain.strain }})
                    </td>
                    <td>
                      <label>
                        <input type="checkbox" class="track-select isotype-item sample-track"
                          value="{{ strain.strain }}" />
                        VCF</label>
                    </td>
                    <td>
                      <label>
                        <input type="checkbox" class="track-select sample-track-alignment"
                          value="{{ strain.strain }}_bam" />
                        BAM</label>
                    </td>
                  </tr>
                  {% endfor %}
                </tbody>
              </table>
            </div>
          </div>
          <!-- /Isotypes -->
        </div>
        <div class="mb-4">
          <label for="track-options" class="form-label">Tracks:</label>
          <!-- Info Modal -->
          <!-- Button trigger modal -->
          <button type="button" class="btn btn-secondary rounded-circle infoButton" data-bs-toggle="modal"
            data-bs-target="#infoModal" aria-label="More information">
            <i class="bi bi-info lh-1" aria-hidden="true"></i>
          </button>
          <!-- Modal -->
          <div class="modal fade" id="infoModal" tabindex="-1" aria-labelledby="infoModalLabel" aria-hidden="true">
            <div class="modal-dialog modal-dialog-scrollable modal-dialog-centered">
              <div class="modal-content">
                <div class="modal-header">
                  <h1 class="modal-title fs-5" id="infoModalLabel">Track Information</h1>
                  <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                </div>
                <div class="modal-body">
                  {% include "_includes/gbrowser-tracks-info.html" %}
                </div>
              </div>
            </div>
          </div>
          <!-- /Info Modal -->
          <!-- Tracks -->
          <div class="p-3 mb-4 rounded-2 overflow-auto gbrowserTracks">
            <form id="track-options">
              {% for i in ["Genes", "Transcripts"] %}
              <div class="checkbox">
                <label><input type="checkbox" checked class="track-select normal-track" value="{{ i }}" /> {{ i }}</label>
              </div>
              {% endfor %}
              {% for i in ["phyloP", "phastCons", "Dust", "Repeat Masker", "Divergent Regions Summary" ] %}
              <div class="checkbox">
                <label><input type="checkbox" class="track-select normal-track" value="{{ i }}" /> {{ i }}</label>
              </div>
              {% endfor %}
              <div class="checkbox">
                <label><input type="checkbox" class="track-select normal-track" value="Divergent Regions" /> Divergent
                  Regions
                  Strain<br /> (<a href="https://andersenlab.org/publications/2021LeeNatureEE.pdf" target='_blank'>Lee
                    <em>et
                      al.</em>
                    2020)</a></label>
              </div>
              <div class="checkbox">
                <label><input type="checkbox" class="track-select normal-track" value="Transposons" /> Transposons (<a
                    href="https://andersenlab.org/publications/2017Laricchia.pdf" target='_blank'>Laricchia <em>et
                      al.</em>
                    2017)</a></label>
              </div>
            </form>
          </div>
          <!-- /Tracks -->
        </div>
        <small>
          <p>IGC (Data Release: {{ release_version }} | Wormbase Version: {{wormbase_version}})</p>
        </small>
      </form>
      <!-- /Input Form -->
    </div>
    <!-- /Input Card -->
  <!-- Tool column -->
  <div class="d-flex flex-wrap flex-column col-md-12 col-xl-8">
      <!-- Options Toolbar -->
      <div class="d-flex flex-wrap justify-content-end col-12 text-end mb-3 optionsToolbar">
      <div class="ps-3 pt-0"><a href="#"><i class="bi bi-arrows-fullscreen" aria-hidden="true"></i> Fullscreen</a></div>
      </div>
      <!-- /Options Toolbar -->
      <!-- IGV Browser -->
      <div class="col">
          <div id="browser"></div>
      </div>
      <!-- IGV Browser -->
  </div>
  <!-- /Tool Column -->
 </div>
 <!-- /Flex -->
</div> <!-- /Container -->

{% endblock %}

{% block script %}


<script type="text/javascript">

var tracks = [];
var trackset = {"Genes": {
                  name: "Genes",
                  displayMode: "EXPANDED",
                  order: 1,
                  url: "{{ track_url_prefix }}/elegans_gene.{{ wormbase_version }}.bed.gz",
                  indexed: false,
                  searchable: false,
                  color: "#5c5cd6",
                  height: 65
                },
                "Transcripts": {
                  name: "Transcripts",
                  url: "{{ track_url_prefix }}/elegans_transcripts_{{ wormbase_version }}.bed.gz",
                  indexed: true,
                  order: 2,
                  color: "#a366ff",
                  displayMode: "SQUISHED",
                  searchable: false
                },
              {% for strain in strain_listing %}
                {{ strain.strain }} : {
                  name: "{{ strain.strain }}",
                  url: "{{ dataset_release_prefix }}/{{ release_version }}/strain/vcf/{{ strain.strain }}.{{ release_version }}.vcf.gz",
                  indexURL: "{{ dataset_release_prefix }}/{{ release_version }}/strain/vcf/{{ strain.strain }}.{{ release_version }}.vcf.gz.tbi",
                  order: 100,
                  displayMode: "EXPANDED",
                  color: "#ffffff",
                  homvarColor: "#0066ff",
                  homrefColor: "#c2c2d6",
                  visibilityWindow: 20000,
                  searchable: false
                },
                {{ strain }}_bam : {
                  id: "{{ strain.strain }}_bam",
                  name: "{{ strain.strain }}_bam",
                  url: "{{ bam_bai_url_prefix }}/{{ strain }}.bam",
                  indexURL: "{{ bam_bai_url_prefix }}/{{ strain }}.bam.bai",
                  order: 100,
                  visibilityWindow: 20000,
                  searchable: false
                },
              {% endfor %}
                "phastCons": {
                  name: "phastCons",
                  url: "{{ track_url_prefix }}/elegans.phastcons.bw",
                  order: 6,
                  displayMode: "SQUISHED",
                  color: "#000000",
                  visibilityWindow: 20000
                },
                "phyloP": {
                  name: "phyloP",
                  url: "{{ track_url_prefix }}/elegans.phylop.bw",
                  order: 6,
                  displayMode: "SQUISHED",
                  color: "#000000",
                  visibilityWindow: 20000
                },
                "Transposons": {
                  name: "Transposons",
                  url: "{{ track_url_prefix }}/laricchia2017.tes_cendr.bed.gz",
                  order: 6,
                  displayMode: "EXPANDED",
                  color: "#34CB80",
                  visibilityWindow: 20000000
                },
                "Divergent Regions Summary": {
                  name: "Divergent Regions Summary",
                  url: "{{ track_url_prefix }}/lee2020.divergent_regions_all.bed.gz",
                  order: 6,
                  height: 40,
                  displayMode: "EXPANDED",
                  color: "#CB3466",
                  visibilityWindow: 20000000
                },
                "Divergent Regions": {
                  name: "Divergent Regions",
                  url: "{{ track_url_prefix }}/lee2020.divergent_regions_strain.bed.gz",
                  order: 6,
                  height: 200,
                  displayMode: "EXPANDED",
                  color: "#9B4763",
                  visibilityWindow: 20000000
                },
                "Dust": {
                  name: "Dust",
                  url: "{{ track_url_prefix }}/c_elegans.PRJNA13758.{{ wormbase_version }}.dust.bed.gz",
                  order: 6,
                  displayMode: "SQUISHED",
                  color: "#583E1A",
                  visibilityWindow: 20000000
                },
                "Repeat Masker": {
                  name: "Repeat Masker",
                  url: "{{ track_url_prefix }}/c_elegans.PRJNA13758.{{ wormbase_version }}.repeat_masker.bed.gz",
                  order: 6,
                  displayMode: "SQUISHED",
                  color: "#CB8C34",
                  visibilityWindow: 20000000
                }
};

function reload_tracks() {
  $('.track-select').each(function(i, obj) {
    const track_name = $(this).attr("value");
    if ($(this).prop("checked") == true){
      if (!tracks.includes(track_name)) {
        tracks.push(track_name);
        igv.getBrowser().loadTrack(trackset[track_name]);
      }
    } else {
      igv.getBrowser().removeTrackByName(track_name);
      i = tracks.indexOf(track_name);
      while(i != -1) {
        tracks.splice(i, 1);
        i = tracks.indexOf(track_name);
      }
    }                          
  });
}

$(document).ready(function () {
  var div = $("#browser")[0], 
      options = {
        search: {
          url: "{{ url_for('api_gene.api_search_combined', query='') }}$FEATURE$",
          coords: 1,
          resultsField: 'result'
        },
        showNavigation: true,
        showKaryo: false,
        reference: {
          id: "{{ wormbase_version }}",
          fastaURL: "{{ track_url_prefix }}/c_elegans.PRJNA13758.{{ wormbase_version }}.genomic.fa",
        },
        locus: "{{ region }}",
        tracks: [],
      };
      var browser = igv.createBrowser(div, options)
                      .then(function(browser) {
                        /*
                          Bind browser events
                        */
                        browser.on('trackdragend', function(reference_frame, label) {
                        });

                        // Detect track changes
                        $(".track-select").on("change", function() {
                          reload_tracks();
                        });

                      });


function process_gene_search() {
    $("#loading-search-table").fadeOut();
    var gene = $('#gene-search').val();
    if (gene.length == 0) {
      $("#g-search-table").fadeOut();
    } else {
      $("#orthologs").html("");
      $.ajax({
           url: "{{ url_for('api_gene.api_search_combined', query='') }}" + gene,
           method: "GET",
           contentType: 'application/xml',
           }).done(function(msg) {
            row = Array();
            $.each(msg, function(i,row) {
                if ("chrom" in row) {
                   link = row["chrom"] + ":" + row["start"] + "-" + row["end"];
                } else {
                    link = row["locus"];
                }
                gene_name = `<a onclick="set_position('${link}')" link='${link}' class='ortholink'>${row["locus"]}</a>`;
                homolog_species = row["homolog_species"] || "<em>C. elegans</em>";
                homolog_source = row["homolog_source"] || "Wormbase";
                result = [gene_name, row['homolog_gene'] || row["gene_symbol"], homolog_species, row["homolog_source"] || "Wormbase"]  ;
                result = "<tr><td>" + result.join("</td><td>") + "</td></tr>";
                position = row["chrom"] + ":" + row["start"];
              $("#orthologs").append(result);
            });
            $("#g-search-table").fadeIn();
            });
    }
}

// Make links work!
$(".container-fluid").on("click", ".ortholink", function() {
  igv.getBrowser().search($(this).attr("link"))
});


function gt_label(gt) {
  // Generates a genotype label
  r = ""
  classes = [];
  classes.push(`gt-${gt["GT"]}`);
  FT = gt["FT"].split(";").join(" ");
  classes.push(FT);
  tt = "";
  if (gt['FT'] != ['PASS']) {
      tt = ` data-placement='bottom' title='${FT}' `;
  }
  r += `<div class='label ttop ${classes.join(" ")}' ${tt} >`;
  r += gt["SAMPLE"] + " : " + gt["TGT"];
  r += "</div>";
  return r
}

function draw_gt_set(genotype_set, genotype_val) {
    return genotype_set.filter(function(gt) { return gt['GT'] == genotype_val })
                       .map( gt_label )
                       .join(" ");

}


var typingTimer;                //timer identifier
var doneTypingInterval = 1000;  //time in ms (5 seconds)

// Ortholog search
$("#gene-search").on("input", function(e) {
  $("#loading-search-table").fadeIn();
  clearTimeout(typingTimer);
  typingTimer = setTimeout(process_gene_search, doneTypingInterval);
})


// Initial load
setTimeout(reload_tracks, 500);

});

set_position = function(coord) {
  // Sets browser position based on input
  igv.getBrowser().search(coord);
}

$(document).ready(function() {

    (function($) {
        var patterns = [];
        $('#filter').keyup(function() {
            $('.searchable tr').hide();
            $(this).val().split(',').forEach(function(r) {
                var rex = new RegExp(r, "i");
                $('.searchable tr').filter(function() {
                    return rex.test($(this).text());
                }).show();
            })
        })

    }(jQuery));


    $('#filter').keydown(function(event) {
        if (event.keyCode == 13) {
            event.preventDefault();
            return false;
        }
    });

});

</script>



{% endblock %}
