{% extends "_layouts/default-nocontainer.html" %}

{% block custom_head %}

<!-- jQuery UI CSS -->
<link rel="stylesheet" type="text/css"
      href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>

<!-- IGV CSS -->

<!-- jQuery JS -->
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
<script>$.widget.bridge('uitooltip', $.ui.tooltip);</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha256-KXn5puMvxCw+dAYznun+drMdG1IFl3agK0p/pqT9KAo=" crossorigin="anonymous"></script>

<script type="text/javascript" src="//igv.org/web/release/2.6.2/dist/igv.min.js"></script>

{% endblock %}

{% block content %}
{% from "_includes/strain-listing.html" import list_strains %}

<div class="container-fluid px-5">
  <div class="d-flex justify-content-around flex-wrap mt-2 pb-5">
    <!-- Input Card -->
    <div class="col-md-8 col-xl-3 text-bg-light mb-5 p-3 p-md-5 rounded-2 shadow-sm">
      <!-- Input Form -->
      <form>
        <!-- Species -->
        {% with onchange = "update_species" %} {% include "_includes/species-selector.html" %} {% endwith %}
        <!-- /Species -->
        <!-- Gene search -->
        <div class="mb-4">
          <label for="gene-search" class="form-label">Gene Search:</label>
          <div class="dropdown">
            <input type="text" class="form-control dropdown-toggle" id="gene-search" placeholder="For example, trt-1"
              data-bs-toggle="dropdown">
            <div class="dropdown-menu table-responsive p-1 dropdownSelect">
              <div class="text-center p-3" id="loading-search-table" style="display:none;">
                <div class="spinner-border" role="status">
                  <span class="visually-hidden">Loading...</span>
                </div>
              </div>
              <table id="g-search-table" class='table table-hover table-striped' style="display:none;">
                <caption class="visually-hidden">A list of species strains.</caption>
                <thead>
                  <tr>
                    <th>Gene</th>
                    <th>Symbol/ID</th>
                  </tr>
                </thead>
                <tbody id="orthologs"></tbody>
              </table>
            </div>
          </div>
        </div>
        <!-- /Gene Search -->
        <div class="mb-4">
          <!-- Isotypes -->
          <label for="filter" class="form-label">Isotype (Reference Strain):</label>
          <div class="dropdown">
            <input type="text" class="form-control dropdown-toggle" id="filter" placeholder="For example, JU360"
              data-bs-toggle="dropdown">
            <div class="dropdown-menu table-responsive dropdownSelect p-1">
              <table id='browser_strain_list' class="table table-striped table-hover">
                <caption class="visually-hidden">A list of strains with different file type options.</caption>
                <tbody class="searchable">
                  {% call(species, strain) list_strains(species_list, strain_listing) %}
                    <td>
                      {{ strain.isotype }} {{ ("(" + strain.strain + ")") if (strain.isotype != strain.strain) else "" }}
                    </td>
                    <td>
                      <label>
                        <input type="checkbox" class="track-select isotype-item sample-track"
                          value="{{ strain.strain }}" />
                        VCF</label>
                    </td>
                    <td>
                      <label>
                        <input type="checkbox" class="track-select sample-track-alignment"
                          value="{{ strain.strain }}_bam" />
                        BAM</label>
                    </td>
                  {% endcall %}
                </tbody>
              </table>
            </div>
          </div>
          <!-- /Isotypes -->
        </div>
        <div class="mb-4">
          <label for="track-options" class="form-label">Tracks:</label>
          <!-- Info Modal -->
          <!-- Button trigger modal -->
          <button type="button" class="btn btn-secondary rounded-circle infoButton" data-bs-toggle="modal"
            data-bs-target="#infoModal" aria-label="More information">
            <i class="bi bi-info lh-1" aria-hidden="true"></i>
          </button>
          <!-- Modal -->
          <div class="modal fade" id="infoModal" tabindex="-1" aria-labelledby="infoModalLabel" aria-hidden="true">
            <div class="modal-dialog modal-dialog-scrollable modal-dialog-centered">
              <div class="modal-content">
                <div class="modal-header">
                  <h1 class="modal-title fs-5" id="infoModalLabel">Track Information</h1>
                  <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                </div>
                <div class="modal-body">
                  {% include "_includes/gbrowser-tracks-info.html" %}
                </div>
              </div>
            </div>
          </div>
          <!-- /Info Modal -->
          <!-- Tracks -->
          <div class="p-3 mb-4 rounded-2 overflow-auto gbrowserTracks">
            <form id="track-options">
              {% for track in default_tracks %}
                <div id="{{ track.name }}_checkbox_container" class="checkbox default-track-container" style="display: none;">
                  <label>
                    <input id="{{ track.name }}_checkbox" type="checkbox" class="track-select normal-track" value="{{ track.name }}" />
                    {{ track.name }}
                    {% if track['source'] %}
                      <br />(<a class="track-source-link" href="{{ track.source.url }}" target='_blank'>{{ track.source.title }}</a>)
                    {% endif %}
                  </label>
                </div>
              {% endfor %}
            </form>
          </div>
          <!-- /Tracks -->
        </div>
        <small>
          <p>IGC (Data Release: <span id="release_version"></span> | Wormbase Version: <span id="wormbase_version"></span>)</p>
        </small>
      </form>
      <!-- /Input Form -->
    </div>
    <!-- /Input Card -->
  <!-- Tool column -->
  <div class="d-flex flex-wrap flex-column col-md-12 col-xl-8">
      <!-- Options Toolbar -->
      <div class="d-flex flex-wrap justify-content-end col-12 text-end mb-3 optionsToolbar">
      <div class="ps-3 pt-0"><a href="#"><i class="bi bi-arrows-fullscreen" aria-hidden="true"></i> Fullscreen</a></div>
      </div>
      <!-- /Options Toolbar -->
      <!-- IGV Browser -->
      <div class="col">
          <div id="browser"></div>
      </div>
      <!-- IGV Browser -->
  </div>
  <!-- /Tool Column -->
 </div>
 <!-- /Flex -->
</div> <!-- /Container -->

{% endblock %}

{% block script %}
{% from "_includes/macros.html" import unpack_species_list %}
{% from "_includes/strain-listing.html" import filter_strains %}
<script type="text/javascript">


/* Variables */


// Typing
var typingTimer;                //timer identifier
var doneTypingInterval = 1000;  //time in ms (5 seconds)

// Tracks
var tracks   = [];
var trackset = {};

// Current active species
var species = null;

// Set of all species
const species_list = {{ unpack_species_list(species_list, species_fields) }};

// Species specific default tracks & strain-specific tracks
var species_tracks = [];

const species_supported_tracks = {
  {% for species_name, species in species_list.items() %}
    "{{species_name}}": [{% for track in species.browser_tracks %} "{{ track }}", {% endfor %}],
  {% endfor %}
};

// Strain track templates
var templates = {};

// Location for browser
const browser_div = $("#browser")[0];

const default_checked_tracks = [
  {% for track in default_tracks %} {% if track.checked %} "{{ track.name }}", {% endif %} {% endfor %}
];



/* Initialization */


$(document).ready(function () {

  // Fetch the browser track objects
  fetch_trackset().then((data) => {
    trackset  = data.default;
    templates = data.templates;

    // Once we know the trackset exists, update the species & browser
    update_species();
  })

  // Italicize 'et al.' in track sources
  $('.track-source-link').each((i, obj) => {
    obj.innerHTML = obj.innerHTML.replaceAll('et al.', '<em>et al.</em>');
  });

  // Isotype search -- make table responsive
  (function($) {
      var patterns = [];
      $('#filter').keyup(function() {
          species_id = species['name'];
          search_terms = $(this).val().split(',');
          {{ filter_strains("species_id", "search_terms") }}
      })
  }(jQuery));

  // Isotype search -- prevent enter key from reloading page
  $('#filter').keydown(function(event) {
      if (event.keyCode == 13) {
          event.preventDefault();
          return false;
      }
  });

  // Ortholog search
  $("#gene-search").on("input", function(e) {
    $("#loading-search-table").fadeIn();
    clearTimeout(typingTimer);
    typingTimer = setTimeout(process_gene_search, doneTypingInterval);
  })

});


// Get the list of tracks
// Retrieves an object with two fields:
//   - 'default':   the list of tracks that are not specific to any one strain
//   - 'templates': the list of track templates to be filled out with strain data
function fetch_trackset() {
  return fetch( "{{ url_for('genome_browser.get_tracks') }}" )
    .then( (response) => { return response.text();  } )
    .then( (text)     => { return JSON.parse(text); } );
}



/* Browser */

function create_or_update_browser(callback) {

  // If browser exists, swap out species params (name and FASTA file)
  if (igv.getBrowser()) {
    return igv.getBrowser().loadGenome({
      "id":       species['name'],
      "name":     species['short_name'],
      "fastaURL": replace_tokens("{{ fasta_filename }}"),
    });
  }

  // If not, create a new browser
  else {

    // Set initial config options for browser
    const options = {
      search: {
        url: "{{ url_for('api_gene.api_search_genes', species='', query='') }}$FEATURE$",
        coords: 1,
        resultsField: 'result'
      },
      showNavigation: true,
      showKaryo: false,
      reference: {
        id:       species['name'],
        name:     species['short_name'],
        fastaURL: replace_tokens("{{ fasta_filename }}"),
      },
      locus: "{{ region }}",
      tracks: [],
    };

    // Create browser, make it responsive, and return it in a Promise
    return igv.createBrowser(browser_div, options).then(function(browser) {

      // Bind browser events
      browser.on('trackdragend', function(reference_frame, label) {});

      // Detect track changes
      $(".track-select").on( "change", update_tracks );

      // Make links work!
      $(".container-fluid").on("click", ".ortholink", function() {
        browser.search($(this).attr("link"))
      });

      // Return browser from Promise
      return browser;
    });
  }
}



/* Browser Tracks */


// Add a single track to the browser by name
function add_track(track_name) {
  if (!tracks.includes(track_name)) {
    tracks.push(track_name);
    return igv.getBrowser().loadTrack(get_track(track_name));
  }
}

// Remove a single track from the browser
function remove_track(track_name) {
  if (track_name == '') return;
  i = tracks.indexOf(track_name);
  while (i != -1) {
    tracks.splice(i, 1);
    i = tracks.indexOf(track_name);
  }
  return igv.getBrowser().removeTrackByName(track_name);
}

// Update tracks in the browser to match the checkboxes
function update_tracks() {
  $('.track-select').each(function(i, obj) {
    const track_name = $(this).attr("value");
    if ($(this).prop("checked") == true) {
      add_track(track_name)
    } else {
      remove_track(track_name)
    }
  });
}

// Remove all tracks from the browser
function clear_tracks() {
  $('.track-select').each(function(i, obj) {
    const track_name = $(this).attr("value");
    remove_track(track_name)
  });
  tracks = [];
}



/* Track Checkboxes */

// Update the browser track checkboxes to match a given array of track names
function set_track_checkboxes(arr) {
  $('.track-select').each(function(i, obj) {
    const track_name = $(this).attr("value");
    obj.checked = arr.includes(track_name);
  });
}

// Reset the track checkboxes to the default values
function reset_track_checkboxes() {
  return set_track_checkboxes(default_checked_tracks);
}

// Uncheck all track checkboxes
function clear_track_checkboxes() {
  return set_track_checkboxes([]);
}



/* Changing Species */

// Show only strains matching current species
function update_strains(species_id) {
  {{ filter_strains("species_id") }}
}


// Show/hide tracks to match list of tracks supported by current species
function update_visible_track_checkboxes(species_id) {
  if (!species_id) {
    hide_track_checkboxes();
  }
  else {
    const suffix = "-checkbox-container";
    $(".default-track-container").each((i, obj) => {
      const name = obj.id.substring( 0, obj.id.length - suffix.length );
      obj.style.display = species_supported_tracks[species_id].includes(name) ? 'block' : 'none';
    });
  }
}

function hide_track_checkboxes() {
  $(".default-track-container").hide();
}


// Set up species variable & set to match species selector
function update_species(selector) {

  // Check for selector element with value
  if ( !selector || !selector.value ) {
    update_visible_track_checkboxes();
    return;
  }

  // Get species from selector
  species_id = selector.value;
  species    = species_list[species_id];

  document.getElementById('wormbase_version').innerHTML = species['wb_ver'];
  document.getElementById('release_version').innerHTML  = species['latest_release'];

  // Update list of strains to match current species
  update_strains(species_id);

  // Reset the checkboxes to the default for the new species
  reset_track_checkboxes(species_id);
  update_visible_track_checkboxes(species_id);

  // Ensure the browser exists and is displaying the correct species,
  // then sort out the tracks in it
  return create_or_update_browser().then((browser) => {

    // Clear any existing tracks from the browser, once we know it exists
    // This is important for default tracks, which are different for different species
    // but have the same track_name.
    clear_tracks();

    // Update the tracks in the browser
    update_tracks();
  });
}



/* Searching */


function process_gene_search() {
  $("#loading-search-table").fadeOut();
  var gene = $('#gene-search').val();
  if (gene.length == 0) {
    $("#g-search-table").fadeOut();
  } else {
    $("#orthologs").html("");
    $.ajax({
        url:         "{{ url_for('api_gene.api_search_genes', query='') }}" + gene + '?species=' + species.name,
        method:      "GET",
        contentType: "application/xml",
    }).done(function(msg) {
        row = Array();
        $.each(msg, function(i, row) {
            if ("chrom" in row) {
                link = row["chrom"] + ":" + row["start"] + "-" + row["end"];
            } else {
                link = row["locus"];
            }
            gene_name = `<a onclick="set_position('${link}')" link='${link}' class='ortholink'>${row["locus"]}</a>`;
            result = [
                gene_name,
                row["gene_symbol"],
            ];
            result = "<tr><td>" + result.join("</td><td>") + "</td></tr>";
            position = row["chrom"] + ":" + row["start"];
            $("#orthologs").append(result);
        });
        $("#g-search-table").fadeIn();
    });
  }
}

function set_position(coord) {
  // Sets browser position based on input
  igv.getBrowser().search(coord);
}



// Replace string tokens with species values
// TODO: Uses of species name ID in filenames not yet standardized, so we drop the 'c_' prefix for now
function replace_tokens(filename) {
  filename = filename.replaceAll('${SPECIES}', species['name'].substring(2));
  {% for token, field in tokens.items() %}
  filename = filename.replaceAll('${' + '{{ token }}' + '}', species['{{ field }}']);
  {% endfor %}
  return filename;
}

// TODO: Move off of client side?
function get_track(track_name) {

  // Retrieve or generate the track
  // If track is in default set, use it
  if (track_name in trackset) {
    var track = trackset[track_name];
  }

  // If not, generate the track from a template
  else {
    if (track_name.endsWith('bam')) {
      var track = fill_track_template('bam', track_name.substring(0, track_name.length - 4));
    } else {
      var track = fill_track_template('vcf', track_name);
    }
  }

  // Get track params
  var params = track.params;

  // Create the full URL and replace all tokens
  params.url = replace_tokens( params.url );

  // If an index URL is defined, replace all tokens
  if (params.indexURL) {
    params.indexURL = replace_tokens( params.indexURL );
  }

  return params;
}



function fill_track_template(template_name, strain_name) {
  var template = JSON.parse( templates[template_name] );

  // Fill in strain names in top level string fields
  for (var key in template) {
    if (typeof template[key] == "string") {
      template[key] = template[key].replaceAll("${STRAIN}", strain_name);
    }
  }

  // Fill in strain name in params
  for (var key in template['params']) {
    if (typeof template['params'][key] == "string") {
      template['params'][key] = template['params'][key].replaceAll("${STRAIN}", strain_name);
    }
  }

  return template;
}



// Generates a genotype label
function gt_label(gt) {

  // Get class list
  classes = gt["FT"].split(";");
  classes.push(`gt-${gt["GT"]}`);

  // Optional extra parameters
  tt = (gt['FT'] != ['PASS']) ? ` data-placement='bottom' title='${FT}' ` : "";

  // Create & return div string
  return `
    <div class='label ttop ${classes.join(" ")}' ${tt}>
      ${gt["SAMPLE"]} : ${gt["TGT"]}
    </div>
  `;
}

function draw_gt_set(genotype_set, genotype_val) {
  return genotype_set.filter( gt => gt['GT'] == genotype_val ).map( gt_label ).join(" ");
}

</script>



{% endblock %}