{% extends "_layouts/default.html" %}

{% block custom_head %}

<!-- jQuery UI CSS -->
<link rel="stylesheet" type="text/css"
      href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"/>

<!-- IGV CSS -->

<!-- jQuery JS -->
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
<script>$.widget.bridge('uitooltip', $.ui.tooltip);</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha256-KXn5puMvxCw+dAYznun+drMdG1IFl3agK0p/pqT9KAo=" crossorigin="anonymous"></script>

<script type="text/javascript" src="//igv.org/web/release/2.6.2/dist/igv.min.js"></script>

{% endblock %}

{% block style %}
<style>

.label {
  width: 80px;
  height: 18px;
  line-height: 12px;
  margin-bottom: 4px;
  display: inline-block;
}

.gt-3 {
    background-color: white;
    border: 1px dotted black;
    color: black;
}

.gt-0.PASS {
  background-color: rgba(194,194,214,1.0);
  border: 1px solid black;
  color: black;
}

.gt-2.PASS {
  background-color: rgba(0, 102, 255,1.0);
  border: 1px solid black;
  color: white;
}

.gt-0:not(.PASS) {
  background-color: rgba(194,194,214,0.25);
  border: 1px dotted black;
  color: black;
}

.gt-2:not(.PASS) {
  background-color: rgba(0, 102, 255,0.25);
  border: 1px dotted black;
  color: black;
}

.het {
  background-color: #ffff00;
  color: black;
}

.gt_set {
  border-right: 1px dotted #b3b3b3;
}

th {
  white-space: nowrap;
}

#variants {
  font-size: 1.0rem;
}

</style>


{% endblock %}


{% block content %}


<div class="browser-row well">

  <div class="row">
    <div class="col-lg-12">

    <!-- Left column -->
    <div class="col-lg-2 col-md-3 col-sm-12">

      <!-- Species dropdown -->
      <h5>Species</h5>
      <select name="species" id="species-selector" onchange="update_species()">
        {% for s in species_list %}
          <option value="{{s}}">{{species_list[s]["short_name"]}}</option>
        {% endfor %}
      </select>

      <!-- Trackset check boxes -->
      <h5>Tracks</h5>
      {% for name, track in default_tracks.items() %}
        <div class="checkbox">
          <label>
            <input id="{{ name }}_checkbox" type="checkbox" class="track-select normal-track" value="{{ name }}" />
            {{ name }}
            {% if 'source' in track %}
              <br />(<a href="{{ track.source.url }}" target='_blank'>{{ track.source.title }}</a>)
            {% endif %}
          </label>
        </div>
      {% endfor %}
    </div> {# /col-lg-2 #}


    <div class="col-lg-4 col-md-5 col-sm-12 ">

      <h5>Gene Search</h5>
      <input type="text" class="form-control" id="gene-search" style="margin-bottom: 5px;" placeholder="Gene Search (e.g. trt-1)">
      <div id="loading-search-table" style="display:none; margin-top:20px;">
        <div class="spinner">
          <div class="bounce1"></div>
          <div class="bounce2"></div>
          <div class="bounce3"></div>
        </div>
      </div>

      <table id="g-search-table" class='table table-striped' style="display: none;">
        <thead>
          <tr>
            <th><em>C. elegans Gene</em></th>
            <th>Symbol/ID</th>
            <th>Species</th>
            <th>Source</th>
          </tr>
        </thead>
        <tbody id="orthologs"></tbody>
      </table>

    </div> {# /col-lg-4 #}


    {# Right column -- Strain table #}
    <div class="col-lg-4 col-md-4 col-sm-12 ">

      <h5>Isotype (reference strain) <a href="{{ url_for('primary.help_item', filename = 'Genome-Browser') }}#ind-strains"><span class="glyphicon glyphicon-question-sign"></span></a></h5>

      {# Strain table search bar #}
      <form class="form-inline pull-xs-right">
        <input class="form-control" id="filter" type="text" style="width:100%;  margin-bottom: 5px;" placeholder="e.g. JU360" autocomplete="off">
      </form>

      {# Strain table -- filled in by update_strains() #}
      <table id='browser_strain_list' class="table table-striped table-hover table-condensed">
        <tbody id="browser_strain_list_body" class="searchable"></tbody>
      </table>
    </div> {# /col-lg-4 #}
</div>
    

  </div> {# /row #}

  <div class="row">

  
    <div class='col-lg-2 col-lg-offset-9 col-md-4 col-md-offset-8 col-sm-offset-4 col-sm-4 col-xs-8 col-xs-offset-2'>
      <a class="btn nu-alt-btn btn-block" href="{{ url_for('primary.help_item', filename = 'Genome-Browser') }}" role="button">
        <span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> 
        Help
      </a>
    </div>
  </div> {# /row #}
  

</div> {# /container #}


  <div class="row igv-browser-row">
    <div class="col-lg-12 col-md-12">
      <h3 class="center">
        IGC (Data Release: <span id="release_version"></span> | Wormbase Version: <span id="wormbase_version"></span>)
      </h3>
      <div id="browser"></div>
    </div>
  </div>

{% endblock %}

{% block script %}


<script type="text/javascript">


/* Variables */


// Typing
var typingTimer;                //timer identifier
var doneTypingInterval = 1000;  //time in ms (5 seconds)

// Tracks
var tracks   = [];
var trackset = {};

// Current active species
var species = null;

// Parse the list of species into a JS object
var species_list = {}
{% for name, species in species_list.items() %}
  species_list['{{ name }}'] = {
    {% for key, val in dict(species).items() %}
      '{{ key }}': '{{ val }}',
    {% endfor %}
  };
{% endfor %}


var strain_list = [];

var templates = {};

// Location for browser
const browser_div = $("#browser")[0];



/* Initialization */


// Load data
$(document).ready(function () {

  // Load species from dropdown default value
  update_species();

  // Set initial config options for browser
  const options = {
    search: {
      url: "{{ url_for('api_gene.api_search_combined', species='', query='') }}$FEATURE$",
      coords: 1,
      resultsField: 'result'
    },
    showNavigation: true,
    showKaryo: false,
    reference: {
      id: "WS276", // TODO: used to be {wormbase_version}, but there's no longer a single value for this
      fastaURL: replace_tokens("{{ site_prefix }}/{{ release_path }}/{{ fasta_filename }}"),
    },
    locus: "{{ region }}",
    tracks: [],
  };

  // Create browser
  var browser = igv.createBrowser(browser_div, options).then(function(browser) {

    /* Bind browser events */
    browser.on('trackdragend', function(reference_frame, label) {});

    // Detect track changes
    $(".track-select").on( "change", update_tracks );

  });

  // Make links work!
  $(".container-fluid").on("click", ".ortholink", function() {
    igv.getBrowser().search($(this).attr("link"))
  });

  // Ortholog search
  $("#gene-search").on("input", function(e) {
    $("#loading-search-table").fadeIn();
    clearTimeout(typingTimer);
    typingTimer = setTimeout(process_gene_search, doneTypingInterval);
  })

  // Set default checked tracks
  {% for name, track in default_tracks.items() %}
    {% if track.checked %}
      document.getElementById("{{name}}_checkbox").checked = true;
    {% endif %}
  {% endfor %}

  // Load tracks from API, then populate the browser with the default tracks
  fetch_trackset( update = true );

  // Load strains for current species from API, then populate table with results
  fetch_strains( update = true );

  // Load browser track templates
  fetch_templates();

});


// Set up isotype search table
$(document).ready(function () {

  // Isotype search
  (function($) {
      var patterns = [];
      $('#filter').keyup(function() {
          $('.searchable tr').hide();
          $(this).val().split(',').forEach(function(r) {
              var rex = new RegExp(r, "i");
              $('.searchable tr').filter(function() {
                  return rex.test($(this).text());
              }).show();
          })
      })
  }(jQuery));

  // Isotype search -- prevent enter key from reloading page
  $('#filter').keydown(function(event) {
      if (event.keyCode == 13) {
          event.preventDefault();
          return false;
      }
  });

});



/* Fetch functions */


// Get the list of tracks
function fetch_trackset( update = false ) {
  fetch( "{{ url_for('gene_browser.get_tracks') }}" )
    .then( (response) => { return response.text();      } )
    .then( (text)     => { trackset = JSON.parse(text); } )
    .then( ()         => { if (update) update_tracks(); } );
}

function fetch_templates() {
  fetch( "{{ url_for('gene_browser.get_track_templates') }}" )
    .then( (response) => { return response.text();       } )
    .then( (text)     => { templates = JSON.parse(text); } );
}

function fetch_strains( update = false ) {
  fetch( "{{ url_for('gene_browser.get_strains', species='') }}" + species.name )
    .then( (response) => { return response.text();         } )
    .then( (text)     => { strain_list = JSON.parse(text); } )
    .then( ()         => { if (update) update_strains();   } );
}



/* Update functions */

function update_tracks() {
  $('.track-select').each(function(i, obj) {
    const track_name = $(this).attr("value");
    if ($(this).prop("checked") == true){
      if (!tracks.includes(track_name)) {
        tracks.push(track_name);
        igv.getBrowser().loadTrack(get_track(track_name));
      }
    } else {
      igv.getBrowser().removeTrackByName(track_name);
      i = tracks.indexOf(track_name);
      while(i != -1) {
        tracks.splice(i, 1);
        i = tracks.indexOf(track_name);
      }
    }
  });
}

function update_strains() {

  // Clear the current contents of the table
  $("#browser_strain_list_body").html("");

  // Map each strain to a table row
  result = strain_list.map((strain) => `
    <tr>
      <td>
        ${ strain.isotype } ${ (strain.isotype != strain.strain) ? "(" + strain.strain + ")" : "" }
      </td>
      <td><label>
          <input type="checkbox" class="track-select isotype-item sample-track" value="${ strain.strain }" />
          VCF
      </label></td>
      <td><label>
          <input type="checkbox" class="track-select sample-track-alignment" value="${ strain.strain }_bam" />
          BAM
      </label></td>
    </tr>
  `);

  // Combine the table rows into one string and insert into the table body
  $("#browser_strain_list_body").html(result.join(''));

  // Detect track changes
  $(".track-select").on( "change", update_tracks );
}

// Set up species variable & set to match species selector
function update_species() {
  species_id = document.getElementById('species-selector').value;
  species    = species_list[species_id];

  document.getElementById('wormbase_version').innerHTML = species['wormbase_version'];
  document.getElementById('release_version').innerHTML  = species['latest_release'];

  fetch_strains( update = true );
}



/* Searching */


function process_gene_search() {
  $("#loading-search-table").fadeOut();
  var gene = $('#gene-search').val();
  if (gene.length == 0) {
    $("#g-search-table").fadeOut();
  } else {
    $("#orthologs").html("");
    $.ajax({
        url:         "{{ url_for('api_gene.api_search_combined', query='') }}" + gene + '?species=' + species.name,
        method:      "GET",
        contentType: "application/xml",
    }).done(function(msg) {
        row = Array();
        $.each(msg, function(i, row) {
            if ("chrom" in row) {
                link = row["chrom"] + ":" + row["start"] + "-" + row["end"];
            } else {
                link = row["locus"];
            }
            gene_name = `<a onclick="set_position('${link}')" link='${link}' class='ortholink'>${row["locus"]}</a>`;
            result = [
                gene_name,
                row['homolog_gene']    || row["gene_symbol"],
                row["homolog_species"] || `<em>${ species_list[ row["species_name"] ]["short_name"] }</em>`,
                row["homolog_source"]  || "Wormbase",
            ];
            result = "<tr><td>" + result.join("</td><td>") + "</td></tr>";
            position = row["chrom"] + ":" + row["start"];
            $("#orthologs").append(result);
        });
        $("#g-search-table").fadeIn();
    });
  }
}

function set_position(coord) {
  // Sets browser position based on input
  igv.getBrowser().search(coord);
}



// Replace string tokens with species values
// TODO: Uses of species name ID in filenames not yet standardized, so we drop the 'c_' prefix for now
function replace_tokens(filename) {
  filename = filename.replaceAll('$SPECIES', species['name'].substring(2));
  {% for token, field in tokens.items() %}
  filename = filename.replaceAll('{{ token }}', species['{{ field }}']);
  {% endfor %}
  return filename;
}

// TODO: Move off of client side?
function get_track(track_name) {

  // Retrieve or generate the track
  // If track is in default set, use it
  if (track_name in trackset) {
    var track = trackset[track_name];
  }

  // If not, generate the track from a template
  else {
    if (track_name.endsWith('bam')) {
      var track = fill_track_template('bam', track_name.substring(0, track_name.length - 4));
    } else {
      var track = fill_track_template('vcf', track_name);
    }
  }

  // Get track params
  var params = track.params;

  // Get the path to the file in GCP
  var path = track['is_bam'] ? "{{ bam_bai_path }}" : "{{ release_path }}"

  // Create the full URL and replace all tokens
  params.url = replace_tokens(`{{ site_prefix }}/${ path }/${ track.filename }`);

  // Create indexURL field by appending file extension to filename, if applicable
  if (track['index_suffix']) {
    params.indexURL = params.url + track['index_suffix']
  }

  return params;
}



function fill_track_template(template_name, strain_name) {
  var template = JSON.parse( templates[template_name] );

  // Fill in strain name in filename
  template['filename'] = template['filename'].replaceAll("${STRAIN}", strain_name);
  template['filename'] = template['filename'].replaceAll("$STRAIN",   strain_name);

  // Fill in strain name in params
  for (var key in template['params']) {
    if (typeof template['params'][key] == "string") {
      template['params'][key] = template['params'][key].replaceAll("${STRAIN}", strain_name);
      template['params'][key] = template['params'][key].replaceAll("$STRAIN",   strain_name);
    }
  }

  return template;
}



function gt_label(gt) {
  // Generates a genotype label
  r = ""
  classes = [];
  classes.push(`gt-${gt["GT"]}`);
  FT = gt["FT"].split(";").join(" ");
  classes.push(FT);
  tt = "";
  if (gt['FT'] != ['PASS']) {
    tt = ` data-placement='bottom' title='${FT}' `;
  }
  r += `<div class='label ttop ${classes.join(" ")}' ${tt} >`;
  r += gt["SAMPLE"] + " : " + gt["TGT"];
  r += "</div>";
  return r
}

function draw_gt_set(genotype_set, genotype_val) {
  return genotype_set.filter(function(gt) { return gt['GT'] == genotype_val })
                      .map( gt_label )
                      .join(" ");
}

</script>



{% endblock %}
