{% extends "_layouts/default-nocontainer.html" %}

{% block custom_head %}
    <script src="//cdnjs.cloudflare.com/ajax/libs/underscore.js/1.5.2/underscore-min.js"></script>
    <script type="text/javascript" src="//igv.org/web/release/2.6.2/dist/igv.min.js"></script>
    <link rel="stylesheet" href="//cdn.datatables.net/1.10.21/css/jquery.dataTables.min.css" />
    <script type="text/javascript" src="//cdn.datatables.net/1.10.21/js/jquery.dataTables.min.js"></script>
{% endblock %}


{% block content %}
{% from "_includes/macros.html" import render_field %}

<div class="container-fluid px-5">
  <div class="d-flex justify-content-around flex-wrap mt-2 pb-5">
    <div class="col-10 col-md-8 col-xl-3">
      <nav>
        <div class="nav nav-tabs border-0" id="nav-tab" role="tablist">
          <button class="nav-link active" id="nav-home-tab" data-bs-toggle="tab" data-bs-target="#nav-home"
            type="button" role="tab" aria-controls="nav-home" aria-selected="true">New</button>
          <button class="nav-link" id="nav-about-tab" data-bs-toggle="tab" data-bs-target="#nav-about" type="button"
            role="tab" aria-controls="nav-about" aria-selected="false">About</button>
            <button class="nav-link" id="nav-linkTwo-tab" type="button" role="tab" aria-controls="nav-linkTwo" aria-selected="false"><a href="{{ url_for('data_releases.data_releases') }}">Release Notes</a></button>
        </div>
      </nav>
      <div class="tab-content" id="nav-tabContent">
        <div class="tab-pane fade show active" id="nav-home" role="tabpanel" aria-labelledby="nav-home-tab"
          tabindex="0">
          <!-- Input Card -->
          <div class="text-bg-light mb-5 p-3 p-md-5 rounded-2 shadow-sm">
            <!-- Input Form -->
            <form id='form-submit' method="POST" onchange="form_change()">
              {{ form.csrf_token }}
              <div class='mb-4'>{{ render_field(form.species,                 onchange="update_species") }}</div>
              <div class="mb-4">{{ render_field(form.strain_1,   disabled='', onchange="update_browser_tracks()") }}</div>
              <div class="mb-4">{{ render_field(form.strain_2,   disabled='', onchange="update_browser_tracks()") }}</div>
              <div class="mb-4">{{ render_field(form.chromosome, disabled='') }}</div>
              <div class="mb-4">{{ render_field(form.start,      disabled='', placeholder="2,028,824") }}</div>
              <div class="mb-4">{{ render_field(form.stop,       disabled='', placeholder="2,029,217") }}</div>
              <div class="d-grid gap-3 mb-4">
                <button id='find_indels' type='submit' value="submit" class="btn btn-secondary text-light" disabled>Find Indels</button>
              </div>
              </form>
              <!-- /Input Form -->
              {% include "_includes/ignore-cache-checkbox.html" %}
              </div>
              <!-- /Input Card -->
        </div>

        <div class="tab-pane fade" id="nav-about" role="tabpanel" aria-labelledby="nav-about-tab" tabindex="0">
          <!-- About Card -->
          <div class="text-bg-light mb-5 p-3 p-md-5 rounded-2 shadow-sm">
            {% include "_includes/pairwise-about-info.html" %}
          </div>
          <!-- /About Card -->
        </div>
      </div>
    </div>
  <!-- Tool column -->
  <div class="d-flex flex-wrap flex-column col-md-12 col-xl-8">
      <!-- Options Toolbar -->
      <div class="d-flex flex-wrap justify-content-end col-12 text-end mb-3 optionsToolbar">
      <div class="ps-3 pt-0"><a href='{{ url_for("pairwise_indel_finder.user_results") }}'><i class="bi bi-journal" aria-hidden="true"></i> My Primer Reports</a></div>
      <div class="ps-3 pt-0"><a href="#"><i class="bi bi-arrows-fullscreen" aria-hidden="true"></i> Fullscreen</a></div>
      </div>
      <!-- /Options Toolbar -->
      <!-- Select Species Prompt -->
      {% include '_includes/select-species-prompt.html' %}
      <!-- /Select Species Prompt -->
      <!-- Browser -->
      <div id="browser-container" class="col">
        <h2>Indel Finder</h2>
        <div id="browser" class="mt-3 mb-5"></div>
        <h2>Primers</h2>
        <table id="results" class='table table-hover table-striped strain-table'></table>
                          <!-- Modal -->
                          <div class="modal fade" id="generate-primers-modal" tabindex="-1" aria-labelledby="generate-primers-modal-label"
                          aria-hidden="true">
                          <div class="modal-dialog modal-dialog-scrollable modal-dialog-centered">
                            <div class="modal-content">
                              <div class="modal-header">
                                <h1 class="modal-title fs-5" id="generate-primers-modal-label">Thank You</h1>
                                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                              </div>
                              <div class="modal-body">
                                <p><i class="bi bi-clock-fill me-1 text-secondary fs-3"></i> Your report will be complete in about
                                  5-10 minutes</li>
                                </p>
                                <p><i class="bi bi-envelope-fill me-1 text-secondary fs-3"></i>You will receive an email when your
                                  report is complete.</p>
                                <p><i class="bi bi-journal me-1 text-secondary fs-3"></i>You may also check <a
                                    href='{{ url_for("pairwise_indel_finder.user_results") }}'>My
                                    Primer Results</a> for updates.</p>
                              </div>
                            </div>
                          </div>
                        </div>
                        <!-- /Modal -->
      </div>
      <!--/Browser -->
  </div>
  <!-- /Tool Column -->
 </div>
 <!-- /Flex -->
</div> <!-- /Container -->

{% endblock %}

{% block script %}
{% from "_includes/strain-listing.html" import unpack_species_list, replace_tokens %}
<script src="{{ url_for('static', filename='js/strains.js') }}"></script>

<script>
{% include '_scripts/utils.js' %}       {#/* defines: fetch_json, toggle_input */#}
{% include '_scripts/igv-browser.js' %}

/* Species */
const species_list = {{ unpack_species_list(species_list, species_fields) }};
var species;

/* Strains */
var strains;

/* IGV Browser */
const browser_div = $("#browser");
const browser_options = {}
var divergent_track_name;
const COLORS = {
  INS: 'blue',
  DEL: 'red'
}

/* Form */
var current_form_data = {
  chromosome: '',
  strain_1:   '',
  strain_2:   '',
}

/* Data Table */
var table;

/* Status */
var query_running = false
var loaded = false



$(document).ready(function(){

  var form = document.getElementById('form-submit');

  species = species_list[$("#{{ form.species.elementId }}").val()];
  update_species();

  // Prevent form submission from reloading page, then redirect to form change logic
  $('form').submit(function (e) {
    e.preventDefault();
    form_change();
  });

	$.ajaxSetup({
		beforeSend: function(xhr, settings) {
			if (!/^(GET|HEAD|OPTIONS|TRACE)$/i.test(settings.type) && !this.crossDomain) {
				xhr.setRequestHeader("X-CSRFToken", "{{ form.csrf_token._value() }}")
			}
		}
	});

  fetch_json("{{ url_for('pairwise_indel_finder.get_strains') }}").then((data) => {
    strains = data;
  });
});


// Query the database for indel data based on the currently selected species, strains, & chromosome
// TODO: Do we have to worry about multiple queries running at once? Will we need to invalidate an older one?
function run_query() {
  query_running = true;
  $.ajax({
    type: "POST",
    url:  "{{ url_for('pairwise_indel_finder.query') }}",
    data: $('form').serialize(),
    success: function(data) {
      results = data["results"]

      // Add the 'generate primers' HTML value to each result
      _.each(results, function(x) {
        x['generate_primers'] = `<div site="${x['CHROM']}:${x["START"]}-${x['END']}" svtype="${x['SVTYPE']}" size="${x["END"]-x["START"]}" class='btn-link generate-primers'>Generate Primers</div>`
      })

      // Handle errors
      $(".form-error").remove();
      if (Object.keys(data).indexOf("errors") > -1) {
        for(error in data["errors"]) {
          $("[for='" + error + "']").parent().append(`<div class='form-error'>${data["errors"][error]}</div>`)
        }
      }

      // If no errors, update the page based on the retrieved results
      else {
        process_indel_finder_query_results(results);
      }

      query_running = false;
    },
  });
}



function form_change() {

  // Ensure browser is loaded
  if (!loaded) return;
  const browser = igv.getBrowser()

  // Get the new values from the form elements
  const species_id = $("#{{ form.species.elementId }}").val().trim();
  const chromosome = $("#chromosome").val().trim();
  const strain_1 = $("#strain_1").val();
  const strain_2 = $("#strain_2").val();

  // Variable to track whether a new query is needed
  // Will be true if chromosome, strain_1, or strain_2 has changed
  var needs_update = false;

  // Enable / disable the start & stop input fields based on whether a species
  // and chromosome have been chosen
  toggle_input('start', species_id && chromosome);
  toggle_input('stop',  species_id && chromosome);

  // Update strain values
  if (current_form_data.strain_1 !== strain_1) {
    current_form_data.strain_1 = strain_1;
    needs_update = true;
  }
  if (current_form_data.strain_2 !== strain_2) {
    current_form_data.strain_2 = strain_2;
    needs_update = true;
  }

  // Update chromosome, resetting start & stop if value has changed
  if (current_form_data.chromosome != chromosome) {
    current_form_data.chromosome = chromosome;
    needs_update = true;
    $('#start').val(null);
    $('#stop').val(null);
  }

  // Update position and wait until new locus values are computed
  // Need browser to compute start & end, in case e.g. range is invalid or the whole chromosome is selected
  set_position(browser).then(({chromosome, start, end}) => {

    // If a new query is needed (chromosome or strain change), run now that position updated
    // Run query if one chromosome and at least one strain are chosen
    if (needs_update && (chromosome !== 'all' && (strain_1 || strain_2))) {
      run_query();
    }
  });
}


function update_species(species_id) {

  const species_empty = species_id == null || species_id == '';

  // If no species selected, prompt user to select one; otherwise show the IGV browser
  $('#select-species-prompt').toggle(species_empty);
  $('#browser-container').toggle(!species_empty);

  // Enable or disable form inputs that depend on species being defined
  const form_ids = [
    'strain_1',
    'strain_2',
    'chromosome',
    'find_indels',
  ];
  form_ids.forEach((form_id) => toggle_input(form_id, !species_empty));

  // If no species chosen, end the update function here
  if (species_empty) return;

  // Set species object from ID
  species = species_list[species_id];

  // Update the browser species & the strain dropdown lists
  create_or_update_browser(browser_div, browser_options, species).then(({browser}) => {
    Promise.all([
      update_strain_dropdown("strain_1", strains, species, ""),
      update_strain_dropdown("strain_2", strains, species, ""),
    ])
    .then(() => {

      // (Re)set the browser position, since it's removed when reference genome changes
      // TODO: Is this the desired behavior? Or should it reset with species change?
      set_position(browser);

      // Make browser responsive
      // If locus is changed from within the browser, update the form values
      // and run a new query if (and only if) the chromosome value has changed
      // Have to be careful here not to create a feedback loop between browser & form
      browser.on('locuschange', function(reference_frame, label) {
        const chrom_old = $("#chromosome").val();
        const chrom_new = reference_frame["chr"];
        $("#chromosome").val(chrom_new);
        $("#start").val(reference_frame["start"][0] === '-' ? 1 : reference_frame["start"]);
        $("#stop").val(reference_frame["end"]);
        if (chrom_old !== chrom_new) {
          run_query();
        }
      });

      // Flag the browser as loaded
      loaded = true;

      // Clear all tracks & reload the divergent regions track
      clear_tracks();
      fetch_json("{{ url_for('pairwise_indel_finder.get_tracks') }}").then((divergent_track_summary) => {
        divergent_track_name = divergent_track_summary.name;

        // Replace species tokens in URL
        divergent_track_summary.url = replace_tokens(divergent_track_summary.url);

        // Set tool-specific display properties
        divergent_track_summary.height = 40; // Height to view two rows would be 70
        divergent_track_summary.removable = false;

        // Add the track to the browser
        add_track(divergent_track_name, divergent_track_summary);
      })

      // Update browser tracks
      update_browser_tracks(browser, results);
    });
  });
}


function process_indel_finder_query_results(results) {

  // Ensure browser already exists
  const browser = igv.getBrowser();
  if (!browser) return;

  // Handle data table
  create_or_update_data_table(results);

  // Handle genome browser
  update_browser_tracks(browser, results);
}



/* Genome Browser Tracks */

// Load the individual strain tracks
function update_browser_tracks(browser, results) {

  // Clear existing strain tracks
  clear_tracks_except([divergent_track_name]);

  // Load new tracks
  load_tracks(browser, results);
}


function load_tracks(browser, results) {
    rowset = {}

    // Create a new array in `rowset` for each selected strain
    if ($("#strain_1").val()) {
      rowset[$("#strain_1").val()] = []
    }
    if ($("#strain_2").val()) {
      rowset[$("#strain_2").val()] = []
    }

    // Restructure results by strain
    for (idx in results) {
      row = results[idx]
      if (rowset[row["STRAIN"]]) {
        rowset[row["STRAIN"]].push(row)
      }
    }

    // Map the sets of track features to single track objects
    // Add the result features to each track
    for (strain in rowset) {
      track = rowset[strain]
      update_track(strain, {
        name: strain,
        type: "annotation",
        features: track.map(row => {
          return {
            chr:   row["CHROM"],
            start: row["START"],
            end:   row["END"],
            name:  row["site"],
            color: COLORS[row["SVTYPE"]],
          };
        }),
      });
    };
  }



/* Data Table Create & Update */


function create_or_update_data_table(results) {

  if (!table) {
    table = $("#results").DataTable({
      data:      results,
      paging:    false,
      ordering:  true,
      searching: false,
      order:     [[1, "asc"]],
      columnDefs: [
        { targets: [0], orderData: [1,2] },
      ],
      columns: [
        { data: "site",   title: "Site"                  },
        { data: "START",  title: "Start", visible: false },
        { data: "END",    title: "End",   visible: false },
        { data: "STRAIN", title: "Strain"                },
        { data: "generate_primers", renderer: "html"     },
      ],
    });
  }
  else {
    table.clear()
    table.rows.add(results);
    table.draw();
  }

  // Activate primer generation links
  $(".generate-primers").on("click", generate_primers);
}


function generate_primers(e) {
  console.log("Submitted")
  $(this).text("Submitted")

  // Get strain values and sort - consistent ordering of strains for hashing
  var strains = [$("#strain_1").val(), $("#strain_2").val()].sort()

  var data = {
    site: $(this).attr("site"),
    strain_1: strains[0],
    strain_2: strains[1],
    size: $(this).attr("size"),
    species: $('#species').val(),
    svtype: $(this).attr("svtype")
  }

  $.ajax({
    type: "POST",
    url:
      {%- if session["is_admin"] -%}
        "{{ url_for('pairwise_indel_finder.submit') }}" + ($('#no_cache_checkbox').prop('checked') ? "?nocache=1" : ""),
      {%- else -%}
        "{{ url_for('pairwise_indel_finder.submit') }}",
      {%- endif %}
    data: JSON.stringify(data),
    contentType: "application/json; charset=utf-8",
    dataType: 'json',
    success: function(result) {
        window.open(`pairwise-indel-finder/result/${result.id}`, "_blank")
    }
  })
}


/* Helper Functions */


// Create a search locus for the IGV browser
function create_search_locus(chromosome, start, end) {
  const DEFAULT_SIZE = 1000000;

  // If no chromosome specified, show all chromosomes
  if (chromosome === '' || chromosome === 'all') {
    return 'all';
  }

  // If no end coordinate specified, show entire chromosome
  // TODO: Might want to tweak this behavior a bit
  if (isNaN(end)) {
    return chromosome;
  }

  // If start is invalid, default to 1
  if (isNaN(start) || start < 1) {
    start = 1;
  }

  // If end is invalid, set to default size
  if (end <= start) {
    end = start + DEFAULT_SIZE;
  }

  return `${chromosome}:${start}-${end}`
}


// Parse an integer form input field to an int
function parse_int_input(selector) {
  return parseInt($(selector).val().trim().replace(/,/g, ''), 10);
}


// Set browser position based on input
// If input position is invalid or incomplete, the final position of the browser
// will be returned by the Promise
function set_position(browser) {
  return new Promise((resolve) => {

    // Extract query information from form
    const chromosome = $("#chromosome").val().trim();
    const start      = parse_int_input("#start");
    const end        = parse_int_input("#stop");

    // Construct search locus string from form info
    const search_locus = create_search_locus(chromosome, start, end)

    // Go to the desired locus, then return it from the Promise
    browser.search(search_locus).then(([locus]) => {
      const chromosome = locus.chromosome.name;
      resolve({
        chromosome,
        start: chromosome === 'all' ? null : locus.start,
        end:   chromosome === 'all' ? null : locus.end,
      });
    });
  })
}


// Replace string tokens with species values
// TODO: Uses of species name ID in filenames not yet standardized, so we drop the 'c_' prefix for now
function replace_tokens(filename) {
  {{ replace_tokens('filename', 'species', tokens) }}
  return filename;
}

</script>

{% endblock %}

